core/controllers/base.py

```diff
@@ -295,8 +300,89 @@ def dispatch(self):
+        except self.InvalidInputException as e:
+            self.handle_exception(e, self.app.debug)
+            schema_validation_succeeded = False
+        # TODO(#13155): Remove this clause once all the handlers have had
+        # schema validation implemented.
+        except NotImplementedError as e:
+            self.handle_exception(e, self.app.debug)
+            schema_validation_succeeded = False
+
+        if not schema_validation_succeeded:
+            return
+
         super(BaseHandler, self).dispatch()
 
+    def validate_args_schema(self):
+        """Validates schema for controller layer handler class arguments.
+
+        Raises:
+            InvalidInputException. Schema validation failed.
+            NotImplementedError. Schema is not provided in handler class.
+        """
+        handler_class_name = self.__class__.__name__
+        request_method = self.request.environ['REQUEST_METHOD']
+        url_path_args = self.request.route_kwargs
+        handler_class_names_with_no_schema = (
+            payload_validator.HANDLER_CLASS_NAMES_WITH_NO_SCHEMA)
+
+        if request_method in ['GET', 'DELETE']:
+            handler_args = {}
+            for arg in self.request.arguments():
+                handler_args[arg] = self.request.get(arg)
+        else:
+            # When request_method in ['PUT', 'POST']
+            handler_args = self.payload
+        if handler_args is None:
+            # When a handler class expects an argument but it is missing in
+            # payloads then for raising exception for missing args,
+            # handler_args must be initialized with empty dict.
+            handler_args = {}
+
+        if handler_class_name in handler_class_names_with_no_schema:
+            return
+
+        if self.URL_PATH_ARGS_SCHEMAS is None:
+            raise NotImplementedError(
+                'Missing schema for url path args in %s '
+                'handler class' % (handler_class_name))
+
+        # For html handlers, extra args are allowed (to accommodate
+        # e.g. utm parameters which are not used by the backend but
+        # needed for analytics).
+        extra_args_are_allowed = (
+            self.GET_HANDLER_ERROR_RETURN_TYPE == 'html' and
+            request_method == 'GET')
+
+        schema_for_url_path_args = self.URL_PATH_ARGS_SCHEMAS
+        self.normalized_request, errors = (
+            payload_validator.validate(
+                url_path_args, schema_for_url_path_args, extra_args_are_allowed)
+        )
+        if errors:
+            raise self.InvalidInputException('\n'.join(errors))
+
+        try:
+            schema_for_request_method = self.HANDLER_ARGS_SCHEMAS[
+                request_method]
+        except Exception:
+            raise NotImplementedError(
+                'Missing schema for %s method in %s handler class.' % (
+                    request_method, handler_class_name))
+
+        self.normalized_request, errors = (
```

Comment:
Do not store it as normalized_request if the request method is PUT or POST.

Reply:
I have handled the cases for 'PUT' and 'POST' in the line 382, do I have to update anything here, please confirm.

Reply:
We should not expect `self.normalized_request` & `self.normalized_payload` to be same for POST or PUT request (Simialr to self.request and self.payload)

Reply:
Done, please take a look.